Support for DBCC CHECKIDENT

aadityavermaa opened the pull request:
### Description

This changes adds support for DBCC CHECKIDENT command in Babelfish that allows user to reset the current value of the identity column in a table. Earlier user had to do the same with Setval() command from pg end-point, with this change they can do it directly from T-SQL end-point. As this is the first time a DBCC command is implemented in Babelfish, this change also provides support to syntax for DBCC commands, and the design is kept scalable so that new DBCC commands can be better introduced in future. 

### Issues Resolved
BABEL-3201

### Test Scenarios Covered ###
* **Use case based -**
Yes

* **Boundary conditions -**
Yes

* **Arbitrary inputs -**
Yes

* **Negative test cases -**
Yes

* **Minor version upgrade tests -**
N/A

* **Major version upgrade tests -**
N/A

* **Performance tests -**
N/A

* **Tooling impact -**
N/A

* **Client tests -**
N/A


### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Commit:
Parser support for BABEL-3201

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
change dbcc_command to CHECKIDENT in newly added grammar

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
remove commented code

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
implementation for NORESEED and RESEED option

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
remove not required comments

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
handle cases where table has no values, added comments

Signed-off-by: Aditya Verma <adiityav@amazon.com>

kuntalghosh commented on contrib/babelfishpg_tsql/src/iterative_exec.c:
```diff
@@ -828,6 +828,15 @@ dispatch_stmt(PLtsql_execstate *estate, PLtsql_stmt *stmt)
 			exec_stmt_restore_ctx_partial(estate, (PLtsql_stmt_restore_ctx_partial *) stmt);
 			break;
+		case PLTSQL_STMT_DBCC:
+			if (pltsql_explain_only)
+				{
+				ereport(ERROR,
+						(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						 errmsg("Showing Estimated Execution Plan for DBCC statment is not yet supported")));
+				}
```
DBCC is utility command. It doesn't make sense to show an execution plan for the same. What's the SQL Server behavior?

aadityavermaa replied:
Executing DBCC CHECKIDENT after SET SHOWPLAN_ALL ON in SQL Server does not throw and error but most of the fields come out to be NULL, I have removed the condition for now.

kuntalghosh replied:
Sorry, which condition exactly have you removed? What's the Babelfish behavior with your changes?

aadityavermaa replied:
Apologies, I just saw that I hadn't included this file in the commit, I meant I have removed the pltsql_explain_only condition for now.

kuntalghosh commented on contrib/babelfishpg_tsql/src/pl_funcs-2.c:
```diff
@@ -717,6 +723,12 @@ dump_stmt_insert_bulk(PLtsql_stmt_insert_bulk *stmt_insert_bulk)
 }
 
+void
+dump_stmt_dbcc(PLtsql_stmt_dbcc *stmt_dbcc)
```
We should also mention the type of the DBCC statement being dumped.

kuntalghosh commented on contrib/babelfishpg_tsql/src/tsqlUnsupportedFeatureHandler.cpp:
```diff
@@ -1239,6 +1239,29 @@ antlrcpp::Any TsqlUnsupportedFeatureHandlerImpl::visitTransaction_statement(TSql
+antlrcpp::Any TsqlUnsupportedFeatureHandlerImpl::visitDbcc_statement(TSqlParser::Dbcc_statementContext *ctx)
+{
+
+	if (ctx->dbcc_command())
```
We should add some telemetry for DBCC statements. Can be addressed separately.

kuntalghosh commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -2955,6 +2959,225 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
 
+int exec_stmt_dbcc(PLtsql_execstate *estate, PLtsql_stmt_dbcc *stmt)
+{
+	switch (stmt->dbcc_stmt_type)
```
If we don't expect the function to be called for other dbcc statement type, we should add an assert(0) for the default case.
```
case default:
		Assert(0);
  
```

aadityavermaa replied:
Made changes in the latest commits, Thanks.

kuntalghosh commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -2955,6 +2959,225 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+	}
+
+	/* Permission check */
+	if (pg_class_aclcheck(table_oid, GetUserId(), ACL_SELECT | ACL_USAGE) != ACLCHECK_OK)
```
Does this check also cover the sysadmin and dbowner permission?

kuntalghosh commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -2955,6 +2959,225 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+
+	pfree(nsp_name);
+
+	/*
+	 * We need to find out the last value of the identity column in the table
+	 * and last value of the identity sequence to compare and decide on the
+	 * action that needs to be taken.
+	 */
```
It's not clear what CHECKIDENT behavior we're trying to implement here. Also, we shouldn't do the relation scan if customer has already provided the new seed value.

aadityavermaa replied:
Updated the comments and finding the max value of the column where it's required ( not when new_reseed_value ) is already provided.

kuntalghosh commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -2955,6 +2959,225 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+	 * and last value of the identity sequence to compare and decide on the
+	 * action that needs to be taken.
+	 */
+	scan = table_beginscan_catalog(rel, 0, NULL);
```
Why are we using catalog scan API to scan an user relation?

aadityavermaa replied:
Since we have to use MAX() to find out the maximum value, I have removed the catalog scan API.

kuntalghosh commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -2955,6 +2959,225 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+
+	if (HeapTupleIsValid(tuple))
+	{
+		max_identity_value = DatumGetInt64(heap_getattr(tuple, attnum+1,
```
The last value need not be the max value. We should execute the select max(id) from table query and fetch the result.

aadityavermaa replied:
Made the change in the latest commit. Thanks.

kuntalghosh commented on test/JDBC/input/BABEL-3201-vu-verify.sql:
```diff
@@ -0,0 +1,35 @@
+DBCC CHECKIDENT(babel_3201_t1);
```
Please try to add at least one test case for all the errors these changes are introducing.

aadityavermaa replied:
I have added test cases for all the functional scenarios, corner cases ( no rows have been inserted in table, rows are removed using TRUNCATE TABLE, rows are deleted using DELETE TABLE), and negative tests for syntax and datatype bounds.

surendravishnoi commented on test/JDBC/expected/BABEL-3201-vu-verify.out:
```diff
@@ -0,0 +1,59 @@
+DBCC CHECKIDENT(babel_3201_t1);
```
Please also add some tests to validate transactional behaviour.
Running DBCC inside a transaction
DBCC running into error inside transaction
Calling DBCC from procedure/functions (with/without errors)

aadityavermaa replied:
Added tests for transactions and procedures. Please take a look. Thanks.

Commit:
addressed review comments

- addreseed review comments
- negative input parameter showed error, fixed
- updated permission checks
- use max(id) to find current identity value
- updated testcases
- fix crash if reseed is used on empty table
-
Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
updated test cases and moved position of permission check

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
made updates

- added permission related tests
- updated BABEL-UNSUPPORTED tests
- added Relation lock on table

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
Merge branch 'BABEL_3_X_DEV' into babel-3201

Commit:
implement output messages for dbcc checkident

Commit:
Re-implemented lock

kuntalghosh commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3112,6 +3119,297 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+			exec_stmt_dbcc_checkident(stmt);
+			break;
+		default:
+			Assert(0);
```
What is the BBF behavior for other DBCC statement types? If we're not throwing an error already, we can't put this assert here.

skumawat2025 replied:
We are throwing error already in the `tsqlUnsupportedFeatureHandler.cpp` for other dbcc statements. 
[Link to Code](https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/1646/files#diff-da105a520d3a543e68071851f0d373f13f9153b9d416627c00a950ed46e81151:~:text=throw%20PGErrorWrapperException(ERROR%2C%20ERRCODE_FEATURE_NOT_SUPPORTED,getLineAndPos(ctx%2D%3Edbcc_command()))%3B)

kuntalghosh commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3112,6 +3119,297 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+	if(dbcc_stmt.new_reseed_value)
+	{
+		/* If float value is passed as reseed_value, only decimal part is considered.*/
+		token = strtok(dbcc_stmt.new_reseed_value, ".");
```
Do we have a testcase for this?

skumawat2025 replied:
yes

surendravishnoi replied:
can we handle cases like ".123", is it a valid use case?

skumawat2025 replied:
as per t-sql its is valid case. with value .123 and -.123 it will treat it as 0 value. Made changes in the latest commit.

kuntalghosh commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3112,6 +3119,297 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+    db_name = get_cur_db_name();
+    if (dbcc_stmt.db_name)
+    {
+        if (pg_strcasecmp(db_name, dbcc_stmt.db_name) != 0)
```
Let's use pg_strncase and provide NAMEDATALEN as length

skumawat2025 replied:
Done.

kuntalghosh commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3112,6 +3119,297 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+			errmsg("relation \"%s\" does not exist", table_name)));
+	}
+
+	rel = RelationIdGetRelation(table_oid);
+	tupdesc = RelationGetDescr(rel);
+
+	/* Find Identity column in table and associated sequence */
+	for (attnum = 0; attnum < tupdesc->natts; attnum++)
+	{
+		Form_pg_attribute attr = TupleDescAttr(tupdesc, attnum);
+
+		if (attr->attidentity)
+		{
+			attname = NameStr(attr->attname);
+			seqid = getIdentitySequence(table_oid, attnum + 1, false);
+			break;
+		}
+	}
+
+	RelationClose(rel);
```
We already have helper function for this. Let's use the same - get_table_identity().

skumawat2025 replied:
Although we can use the `get_table_identity()` api to find `seqid` but we will also need the name of this sequence  col  that is `attname` in the code  to find maximum value. See line no. 3361.

kuntalghosh commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3112,6 +3119,297 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+	PG_TRY();
+	{
+		// Lock table to ensure concurrency.
+		if(dbcc_stmt.is_reseed)
```
Please add a comment why this is needed

skumawat2025 replied:
Fixed.

kuntalghosh commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3112,6 +3119,297 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+		 */
+		if (cur_value_is_null)
+		{
+			if (dbcc_stmt.new_reseed_value)
+			{
+				snprintf(message, sizeof(message), "Checking identity information: current identity value 'NULL'.\n");
+
+				DirectFunctionCall3(setval3_oid,
+					ObjectIdGetDatum(seqid),
+					Int64GetDatum(reseed_value),
+					BoolGetDatum(false));
+			}
+			else
+			{
+				snprintf(message, sizeof(message), "Checking identity information: current identity value 'NULL', current column value 'NULL'.\n");
+			}
```
Do we have test cases for these two scenarios?

skumawat2025 replied:
Added.

Commit:
Merge branch 'BABEL_3_X_DEV' into babel-3201

kuntalghosh commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3112,6 +3119,297 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+		// Lock table to ensure concurrency.
+		if(dbcc_stmt.is_reseed)
+		{
+			LockRelationOid( table_oid, AccessExclusiveLock);
+		}
+		else
+		{
+			LockRelationOid( table_oid, ShareLock);
```
Please add a concurrency test for this scenario where other backends are blocked from accessing this table while a DBCC CHECKIDENT has reseeded the identity column. You can use isolation test framework.

skumawat2025 replied:
added test for concurrency using isolation test framework.

kuntalghosh commented on test/JDBC/input/BABEL-3201-vu-verify.mix:
```diff
@@ -0,0 +1,247 @@
+SELECT * from babel_3201_t_bigint;
+GO
+
+-- no rows have been inserted in the table; both current identity value current
+-- column value should be NULL (TO-DO)
+DBCC CHECKIDENT(babel_3201_t1, NORESEED);
+GO
+
+DBCC CHECKIDENT(babel_3201_t1, RESEED);
+GO
+
+-- current identity value should be NULL (TO-DO), identity value inserted in
+-- next INSERT operation should be new_reseed_value.
+DBCC CHECKIDENT(babel_3201_t1, RESEED, 5);
+GO
+
+INSERT INTO babel_3201_t_int VALUES (5);
+GO
+
+SELECT * FROM babel_3201_t_int;
+GO
```
The tests don't match the comment. The table is not empty

skumawat2025 replied:
Fixed.

kuntalghosh commented on test/JDBC/input/BABEL-3201-vu-verify.mix:
```diff
@@ -0,0 +1,247 @@
+TRUNCATE TABLE babel_3201_t_int;
+GO
+
+DBCC CHECKIDENT(babel_3201_t_int, RESEED, 10);
```
Also, add a test with noreseed.

skumawat2025 replied:
added.

kuntalghosh commented on test/JDBC/input/BABEL-3201-vu-verify.mix:
```diff
@@ -0,0 +1,247 @@
+ROLLBACK;
+GO
+
+SELECT * FROM babel_3201_t_decimal;
```
For ROLLBACK, we need to check what's the current identity value? In PG, it'll not be rolled-back - different from SQL Server.

skumawat2025 replied:
Added a `checkident(<table>, noreseed)` after transection rollback to check the current identity values, which is not rolled back.

Commit:
This commit includes-

- fix for locking behaviour
- tests in isolation test framework for concurrency
- Add some missing jdbc tests
- Address review comments

Signed-off-by: Sandeep Kumawat <skumwt@amazon.com>

Commit:
Merge branch 'BABEL_3_X_DEV' into babel-3201

Commit:
Fix: test failures

Signed-off-by: Sandeep Kumawat <skumwt@amazon.com>

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3112,6 +3119,309 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+    db_name = get_cur_db_name();
+    if (dbcc_stmt.db_name)
+    {
+        if (pg_strncasecmp(db_name, dbcc_stmt.db_name, NAMEDATALEN) != 0)
```
Please add testcase when db_name is more than 64 chars and we are trying to do cross-db call.

skumawat2025 replied:
added. please have a look.

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3112,6 +3119,309 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+		{
+			schema_name = get_authid_user_ext_schema_name((const char *) db_name, user);
+			nsp_name = get_physical_schema_name(db_name, schema_name);
+		}
```
Missing dbo user handling. Please refer set_search_path_for_user_schema() for more details.

skumawat2025 replied:
made changes. Please have a look.

surendravishnoi commented on contrib/babelfishpg_tsql/src/tsqlIface.cpp:
```diff
@@ -1889,6 +1890,22 @@ class tsqlBuilder : public tsqlCommonMutator
+	void enterDbcc_statement(TSqlParser::Dbcc_statementContext *ctx) override
+	{
+		if (ctx->CHECKIDENT())
+        	graft(makeDbccCheckidentStatement(ctx), peekContainer());
```
indentation

surendravishnoi commented on contrib/babelfishpg_tsql/src/tsqlIface.cpp:
```diff
@@ -1889,6 +1890,22 @@ class tsqlBuilder : public tsqlCommonMutator
+        	graft(makeDbccCheckidentStatement(ctx), peekContainer());
+
+		clear_rewritten_query_fragment();
+		// PLtsql_stmt_execsql *stmt = (PLtsql_stmt_execsql *) getPLtsql_fragment(ctx);
```
cleanup if not required

surendravishnoi commented on contrib/babelfishpg_tsql/src/tsqlIface.cpp:
```diff
@@ -5416,6 +5432,72 @@ makeExecBodyBatch(TSqlParser::Execute_body_batchContext *ctx)
+		if (ctx->table_name()->schema)
+			schema_name = stripQuoteFromId(ctx->table_name()->schema);
+		if (ctx->table_name()->table)
+		table_name = stripQuoteFromId(ctx->table_name()->table);
```
indentaion

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3112,6 +3119,309 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+	int64	max_identity_value = 0;
+	int64	cur_identity_value = 0;
+	char	*max_identity_value_str = NULL;
+    char	*login = GetUserNameFromId(GetSessionUserId(), false);
```
indentation

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3112,6 +3119,309 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+	char		message[200];
+
+
+    bool    is_cross_db = false;
```
indentation

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3112,6 +3119,309 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+    bool    is_cross_db = false;
+	if(dbcc_stmt.new_reseed_value)
+	{
+		/* If float value is passed as reseed_value, only decimal part is considered.*/
```
/* comment */

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3112,6 +3119,309 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+    {
+        if (pg_strncasecmp(db_name, dbcc_stmt.db_name, NAMEDATALEN) != 0)
+            is_cross_db = true;
+        db_name = pstrdup(dbcc_stmt.db_name);
```
do we need to free earlier db_name?

skumawat2025 replied:
made changes.

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3112,6 +3119,309 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+        db_name = pstrdup(dbcc_stmt.db_name);
+    }
+
+	if (!DbidIsValid(get_db_id(db_name)))
```
indentation is messed up at number of places

surendravishnoi commented on test/JDBC/expected/BABEL-3201-vu-prepare.out:
```diff
@@ -0,0 +1,146 @@
+~~ROW COUNT: 1~~
+
+
+CREATE PROCEDURE babel_3201_proc1
```
can we call DBCC from a function/view?

skumawat2025 replied:
We can call DBCC inside a transaction and proc but not inside a function. test-cases for proc/tran are there.

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3112,6 +3119,309 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+        if (user)
+            SetCurrentRoleId(GetSessionUserId(), false);
+        else
+            ereport(ERROR,
```
please add a test case for this error if not already

skumawat2025 replied:
added.

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3112,6 +3119,309 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+		 */
+		if (!user)
+		{
+			ereport(ERROR,
```
please add a test case

skumawat2025 replied:
I think adding test-case for this error is not possible. `get_user_for_database(db_name)` will check if the login is the db owner, or login is member of sysadmin, or guest is enabled in login. Otherwise it will return NULL. This case we have covered under corss_db calls. And in cross-db calls when user is NULL we will throw error `The server principal \"%s\" is not able to access the database \"%s\" under the current security context`

surendravishnoi replied:
not clear. can you test without cross db? also, if we cannot write a test case means that code path is not possible. If that is the case, add an assert (at-least for debug)

skumawat2025 replied:
No, user can  not be null without cross-db calls.  
For example we can see that api  `get_user_for_database(db_name)` will return NULL when we are trying to do a cross db call and there is no user present for that db for the current login. Adding an assert for debugging.

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3112,6 +3119,309 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+		}
+		
+		if (is_cross_db)
+            SetCurrentRoleId(current_user_id, false);
```
identation

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3112,6 +3119,309 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+	{
+		PG_RE_THROW();
+	    if (is_cross_db)
+            SetCurrentRoleId(current_user_id, false);
```
control will never come here, we should first do set and then rethrow

skumawat2025 replied:
made changes. please have a look.

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3112,6 +3119,309 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+	}
+	PG_CATCH();
+	{
+		PG_RE_THROW();
```
who will release lock in case of errors?

skumawat2025 replied:
made changes to release locks in case of errors.

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3112,6 +3119,309 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+
+	if(!dbcc_stmt.new_reseed_value)
+	{
+		UnlockRelationOid(table_oid, ShareLock);
```
what about exclusive lock?

skumawat2025 replied:
So, we are trying to hold this shareLock on table only for the execution of this statement. We try to hold Exclusive Lock on a transaction level. We want to release this share lock as soon as we are done with the execution of DBCC CHECKIDENT() stmt. But Exclusive Lock will get release when transection gets commited or rollbacked.

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3112,6 +3119,309 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+	}
+	PG_END_TRY();
+	
+	if (max_identity_value_str)
```
we want to free memory for error cases as well

skumawat2025 replied:
done. made changes in the next commit. please take a look.

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3112,6 +3119,309 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+		if (is_cross_db)
+            SetCurrentRoleId(current_user_id, false);
+	}
+	PG_CATCH();
```
do we have test cases with different error and lock types?

skumawat2025 replied:
So there are test-cases in isolation test framwork, when one transection made update to table t1 (by reseeding identity value via dbcc checkident) and other transection is not able to use t1 inside a dbcc checkident stmt.

Commit:
This commit includes

- Correct Parser Support for DBCC CHECKIDENT,<table_name> in quotes is the correct way.
- Update test files accordingly
- Adds more test-cases for cross-db errors, db_name longer then 63,
  Mixed-cases names, etc.
- Made updates to free memory in case of errors

Signed-off-by: Sandeep Kumawat <skumwt@amazon.com>

Commit:
Merge branch 'BABEL_3_X_DEV' into babel-3201

HarshLunagariya commented on contrib/babelfishpg_tsql/antlr/TSqlParser.g4:
```diff
@@ -4271,15 +4305,21 @@ keyword
     | CHANGE_TRACKING
+    | CHECKDB
+    | CHECKFILEGROUP
+    | CHECKIDENT
```
Do we still need these changes ? since in the dbcc_command rule we have already explicitly specified them.

skumawat2025 replied:
Correct. Removing.

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -19,6 +20,8 @@
 #include "pl_explain.h"
 #include "session.h"
 
+#include <math.h>
```
Is it needed?
I see we are using pg_strtoint64 which is there 'utils/builtins.h' which is indirectly included by some other header.

skumawat2025 replied:
Yes. Removing it .

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3113,6 +3120,347 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+		else
+		{
+			nsp_name = get_physical_schema_name(db_name, dbcc_stmt.schema_name);
+		}
```
This is incorrect. Here dbcc_stmt will be NULL We are missing coverage for this case otherwise it would have failed.
It should be schema_name instead. Move schema_name evaluation inside this condition.

skumawat2025 replied:
Thanks for pointing this out. Fixing it and also adding a test-case for the coverage. However as we do need schema_ name later ahead in the code, keeping schema_name evaluation out of this else condition

Commit:
Address review comments

Signed-off-by: Sandeep Kumawat <skumwt@amazon.com>

Commit:
address test failures

Signed-off-by: Sandeep Kumawat <skumwt@amazon.com>

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3113,6 +3118,347 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+
+		if (is_float_value)
+		{
+			if (dbcc_stmt.new_reseed_value[0] == '.' || 
```
have we tried input like ".", "-" etc for testing?

skumawat2025 replied:
tested and added cases.

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3113,6 +3118,347 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+			(errcode(ERRCODE_UNDEFINED_DATABASE),
+				errmsg("database \"%s\" does not exist", dbcc_stmt.db_name)));
+		}
+       		if (pg_strncasecmp(db_name, dbcc_stmt.db_name, NAMEDATALEN) != 0)
```
make sure we have test cases for mixed case db name and DB name more than 64 chars (if allowed)

skumawat2025 replied:
There are test cases with db name longer then 64 char, mixed case db name.

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3113,6 +3118,347 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+    	}
+
+	user = get_user_for_database(db_name);
+	login_is_db_owner = 0 == strncmp(GetUserNameFromId(GetSessionUserId(), false),
```
make sure we have test cases for mixed case user name and user name more than 64 chars (if allowed)

skumawat2025 replied:
added a test case where user name is longer then 64 chars.

Commit:
Merge branch 'BABEL_3_X_DEV' into babel-3201

Commit:
This commit -

-  Address PR comments
-  Fixes locking behaviour

Signed-off-by: Sandeep Kumawat <skumwt@amazon.com>

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3113,6 +3118,359 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+		/* user will never be null here as cross-db calls are already handled */
+		Assert(user != NULL);
+
+		schema_name = get_authid_user_ext_schema_name((const char *) db_name, user);
```
Please move it inside the third condition, It is not needed for every condition. Otherwise permissions related changes looks good.

HarshLunagariya reviewed (commented):
Verified user and permissions related changes. Looks good!
